Add constraint for Language Tree Node

hannaseithe opened the pull request:
### Short description
<!-- Describe this PR in one or two sentences. -->
This adds a constraint to the Language Tree Node model, so that creating a node that is both inactive and visible is not possible anymore. 

### Proposed changes
<!-- Describe this PR in more detail. -->

- the constraint is added in the model
- migrations both for the new constraint and mutating possible nodes with active=False & visible=True to visible=False were added
- a test for the constraint was created


### Side effects
<!-- List all related components that have not been explicitly changed but may be affected by this PR -->

- no side effects


### Faithfulness to issue description and design
<!-- If the implementation is different from the issue description and design, replace the following with an explain why. -->
There are no intended deviations from the issue and design.


### Resolved issues
<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #2133


__________________________________________________
<!-- Keep this link for the potential reviewer -->
[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

hannaseithe commented:
I am wondering if we should also automatically uncheck and disable the checkbox for "visible"  in the form for creating a new language tree node, if "active" is unchecked? @nikolahoff . 

Right now I have implemented it so far, that you are still able to select in the form the combination **active=>unchecked & visible => checked**. But if you try to create such a language tree node it fails and an error warning from the server is shown.

nikolahoff commented:
@hannaseithe yes, as you described the behaviour makes sense to me.

MizukiTemma commented on integreat_cms/locale/de/LC_MESSAGES/django.po:
```diff
@@ -10050,6 +10058,30 @@ msgstr "aktiviert"
+msgid_plural ""
+"The following {model_name} could not be deactivated, because inactive "
+"language tree nodes cannot be visible: {object_names}"
+msgstr[0] ""
+"{model_name} {object_names} konnte nicht deaktiviert werden, da die Quell-"
+"Übersetzung fehlt."
+msgstr[1] ""
+"Die folgenden {model_name} konnten nicht deaktiviert werden, da ihre Quell-"
+"Übersetzungen fehlen: {object_names}"
+
```
Maybe a copy&paste error? The translation for the reason must be adjusted 😅

hannaseithe replied:
Thank you for the CR!

MizukiTemma reviewed (changes requested):
Looks good! There is only two things to be adjusted 🇩🇪

MizukiTemma commented on integreat_cms/cms/templates/languagetreenodes/languagetreenode_form.html:
```diff
@@ -21,7 +21,8 @@ <h1 class="heading overflow-hidden text-ellipsis">
                 <div class="flex flex-wrap gap-2">
                     <a href="{% url 'languagetreenodes' region_slug=request.region.slug %}"
-                       class="btn btn-ghost">{% translate "Cancel" %}</a>
+                       class="btn btn-ghost">{% translate
+                    "Cancel" %}</a>
```
This made the translation broken and resulted in such appearance of the button

![image](https://github.com/user-attachments/assets/a2048f71-43c5-4a2c-b286-35db1625da61)


```suggestion
                       class="btn btn-ghost">{% translate "Cancel" %}</a>
```

MizukiTemma reviewed (changes requested):
Thank you for updates 💪 

There is one thing left in the bulk action to fix: making inactive languages visible by bulk action leads to an error.

hannaseithe commented:
> Sorry for requesting changes again and again. I didn't find all points to fix in the last review 🙈
Dont feel bad for being thorough! That way I learn and we get good quality code and releases. I genuinely appreciate it :smile_cat:

JoeyStk reviewed (approved):
Thank you very much for your improvements! To me this looks ready to be merged :rocket:

hannaseithe commented on integreat_cms/cms/migrations/0135_languagetreenode_language_tree_node_inactive_requires_invisible.py:
```diff
@@ -0,0 +1,32 @@
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("cms", "0134_alter_region_status"),
```
I am very much looking forward to this ... not :D But thank you for the reminder!

JoeyStk replied:
It can be daunting for sure! I hope that no inconveniences come up :crossed_fingers:

MizukiTemma commented:
> Sorry for requesting changes again and again. I didn't find all points to fix in the last review 🙈
> 
> * Migration fails, if there is any language node in the DB that is inactive but vidible, as it can fulfill the new constraint 😿
>   How to reproduce:
> 
> 1. Check out to the branch develop
> 2. ./tools/run.sh and make a language node inactive but visible
> 3. Check out to the branch of this PR
> 4. ./tools/run.sh (without ./tools/prune_database.sh)
> 5. See django.db.utils.IntegrityError
> 
> As we cannot guarantee that there is no such language node at the moment of deployment, I suggest to add a function to deal with such a case.
> 
> [This migration](https://github.com/digitalfabrik/integreat-cms/blob/develop/integreat_cms/cms/migrations/0138_configurable_mt_budget.py) would be an example for that. Write a function that turns all inactive and visible language nodes into inactive and invisible, like `migrate_mt_addon_booked` in the exmaple and apply this before `migration.AddConstraint`. This way it should work (but I didn't try by myself 🙈 sorry, if it won't go well).
> 
> * There was once case found when a bulk action faces an user-unfriendly error:
> 
> 1. Make a language node inactive and invisible
> 2. Make the language node visible by bulk Action
> 3. You'll face this
> 
> <img alt="image" width="1185" height="267" src="https://private-user-images.githubusercontent.com/82345046/467084297-51826ccb-ded1-4216-97eb-9e1397411997.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ0ODk2NzgsIm5iZiI6MTc1NDQ4OTM3OCwicGF0aCI6Ii84MjM0NTA0Ni80NjcwODQyOTctNTE4MjZjY2ItZGVkMS00MjE2LTk3ZWItOWUxMzk3NDExOTk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODA2VDE0MDkzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNhNDUzMDcxM2UwZWVmMGE5OWNkYWU2MjVjZjhjNzQyZGU5OGQ1ZTc0Mjk0ZDQ1ZjFlZDg1NDYzNGRiM2RjMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7q79iz4dyg9nkqNyNcgSP9z4X2kzrFl_uDVsmyGCbRw">

@hannaseithe 
Seems this is not yet fixed? 👀

hannaseithe commented:
@MizukiTemma / @JoeyStk

For Mizuki's change request I modified the BulkUpdateBooleanFieldView's post request to catch the Integrity Error directly and create an error message (I deleted my changes to BulkDisableView's post request as this will also be covered in the super class). When doing so I ran into an TransactionManager Error, which was caused by us wrapping LanguageTreeBulkActionView in an atomic transaction (with the @tree_mutex() decorator). After looking into it I am quite confident that the wrapper is unnecessary here, since we do not call any treebeard methods from this view (we only change the values of boolean fields) - that is why I have deleted the decorator and handle the IntegrityError with a simple try block inside BulkUpdateBooleanFieldView's post method. 

I would kindly ask you both to review the changes.

@PeterNerlich also mentioning you here since we talked about whether the decorator can be deleted or not

PeterNerlich commented:
Just giving my two cents here: Over the past we have learnt to be very cautious of anything related to treebeard and that we rather want to accept some inefficiencies that we are able to just throw more computing resources on than potentially opening up more holes where treebeard could come back to bite us. Although *(or especially because)* the exact mechanism is not understood by me – and you assert that treebeard does not overwrite any of the save, update etc. methods or use signals – we have observed treebeard sporadically performing big changes that go far beyond the item to save itself, silently wreaking havoc when used in a transaction that got rolled back *(but treebeards changes persisting back then)*, possibly involving outdated data in the cache or django memory, or just concurrent treebeard operations. After #3076 this is still not fully resolved *(though my personal hope is that it got less)*.

If we can get away with leaving the decorator in out of caution, I would prefer that. [According to everything I'm reading online](https://stackoverflow.com/a/69127707) it should be possible to roll back only part of an atomic transaction as you do. I don't understand where the `TransactionManagementError` is coming from

hannaseithe commented:
Ok I have implemented it now with the decorator in place. This is the version I came up with together with @JoeyStk. 

There is an alternate version to be found here: https://github.com/digitalfabrik/integreat-cms/tree/enhancement/prevent-inactive-and-visible-ltn-alt . This also keeps the decorator, but sets the savepoint and handles the IntegrityError inside `LanguageTreeBulkActionView`. But after learning that setting a `with transaction.atomic():` context around a single update doesnt make a difference when there is no decorator (which could happen in the future if another class than `LanguageTreeBulkActionView` inherits from `BulkUpdateBooleanView`), I prefer the solution presented here.

Now its really ready for review :)

MizukiTemma commented:
@hannaseithe 

> I prefer the solution presented here.

Which one is it? Does "here" menas  the branch `enhancement/prevent-inactive-and-visible-ltn` or `enhancement/prevent-inactive-and-visible-ltn-alt`?

hannaseithe commented:
> @hannaseithe
> 
> > I prefer the solution presented here.
> 
> Which one is it? Does "here" menas the branch `enhancement/prevent-inactive-and-visible-ltn` or `enhancement/prevent-inactive-and-visible-ltn-alt`?

I meant the branch `enhancement/prevent-inactive-and-visible-ltn` . This is what I prefer
